Quick improvements for filterDescendants

SamyPesse opened the pull request:
Avoid doing concat on immutable `List` and instead build an array then return it.

It also adds a `forEachDescendant` when we just want to iterate over nodes without creating a return list.

SamyPesse commented:
### [With this PR](https://travis-ci.org/GitbookIO/slate/jobs/172898070):

```
Benchmarks
   delete-backward-deep
     Current:	 263 ops/sec (107 runs sampled)
     rme: ±6.36%
   delete-backward-normal
     Current:	 328 ops/sec (109 runs sampled)
     rme: ±6.26%
   delete-backward-wide
     Current:	 340 ops/sec (97 runs sampled)
     rme: ±4.23%
   first-render-normal
     Current:	 64.97 ops/sec (120 runs sampled)
     rme: ±4.22%
   insert-text-normal
     Current:	 926 ops/sec (95 runs sampled)
     rme: ±4.48%
   memoize-util
     Current:	 8,478 ops/sec (86 runs sampled)
     rme: ±8.33%
   normalize-document-deep
     Current:	 1,275 ops/sec (88 runs sampled)
     rme: ±8.98%
   normalize-document-normal
     Current:	 2,671 ops/sec (84 runs sampled)
     rme: ±6.87%
   normalize-document-wide
     Current:	 1,135 ops/sec (83 runs sampled)
     rme: ±4.91%
   render-split-block
     Current:	 27.99 ops/sec (111 runs sampled)
     rme: ±2.81%
   split-block-deep
     Current:	 253 ops/sec (96 runs sampled)
     rme: ±6.42%
   split-block-normal
     Current:	 416 ops/sec (94 runs sampled)
     rme: ±5.76%
   split-block-wide
     Current:	 338 ops/sec (88 runs sampled)
     rme: ±7.75%
```

### [Previously](https://travis-ci.org/ianstormtaylor/slate/jobs/172893921)

```
Benchmarks
   delete-backward-deep
     Current:	 249 ops/sec (107 runs sampled)
     rme: ±4.61%
   delete-backward-normal
     Current:	 318 ops/sec (110 runs sampled)
     rme: ±4.83%
   delete-backward-wide
     Current:	 306 ops/sec (96 runs sampled)
     rme: ±4.73%
   first-render-normal
     Current:	 65.58 ops/sec (123 runs sampled)
     rme: ±3.39%
   insert-text-normal
     Current:	 777 ops/sec (97 runs sampled)
     rme: ±6.27%
   memoize-util
     Current:	 3,409 ops/sec (97 runs sampled)
     rme: ±5.25%
   normalize-document-deep
     Current:	 1,346 ops/sec (90 runs sampled)
     rme: ±5.71%
   normalize-document-normal
     Current:	 3,006 ops/sec (85 runs sampled)
     rme: ±4.99%
   normalize-document-wide
     Current:	 1,320 ops/sec (83 runs sampled)
     rme: ±7.69%
   render-split-block
     Current:	 30.32 ops/sec (104 runs sampled)
     rme: ±6.02%
   split-block-deep
     Current:	 222 ops/sec (101 runs sampled)
     rme: ±4.17%
   split-block-normal
     Current:	 367 ops/sec (95 runs sampled)
     rme: ±7.16%
   split-block-wide
     Current:	 278 ops/sec (92 runs sampled)
     rme: ±5.79%
```

Soreine commented on src/models/node.js:
```diff
@@ -920,11 +935,7 @@ const Node = {
-    return this.nodes.reduce((texts, node) => {
-      return node.kind == 'text'
-        ? texts.push(node)
-        : texts.concat(node.getTexts())
```
`getTexts()` had the advantage of being memoized recursively

SamyPesse commented:
Here is the comparaison between my last 2 commits:

```
Benchmarks

   delete-backward-deep
     Current:	 299 ops/sec (113 runs sampled)
     Reference:	 305 ops/sec (108 runs sampled)
     comparison: Indeterminate
     diff: -1.94%
     rme: ±3.23%

   delete-backward-normal
     Current:	 376 ops/sec (116 runs sampled)
     Reference:	 369 ops/sec (113 runs sampled)
     comparison: Indeterminate
     diff: +2.07%
     rme: ±2.50%

   delete-backward-wide
     Current:	 354 ops/sec (99 runs sampled)
     Reference:	 362 ops/sec (100 runs sampled)
     comparison: Indeterminate
     diff: -2.15%
     rme: ±4.16%

   first-render-normal
     Current:	 83.48 ops/sec (126 runs sampled)
     Reference:	 79.54 ops/sec (124 runs sampled)
     comparison: Faster
     diff: +4.95%
     rme: ±3.23%

   insert-text-normal
     Current:	 1,046 ops/sec (97 runs sampled)
     Reference:	 1,040 ops/sec (96 runs sampled)
     comparison: Faster
     diff: +0.59%
     rme: ±4.54%

   memoize-util
     Current:	 4,849 ops/sec (98 runs sampled)
     Reference:	 4,695 ops/sec (101 runs sampled)
     comparison: Indeterminate
     diff: +3.28%
     rme: ±8.31%

   normalize-document-deep
     Current:	 1,935 ops/sec (89 runs sampled)
     Reference:	 1,996 ops/sec (88 runs sampled)
     comparison: Indeterminate
     diff: -3.06%
     rme: ±5.63%

   normalize-document-normal
     Current:	 3,969 ops/sec (85 runs sampled)
     Reference:	 4,180 ops/sec (86 runs sampled)
     comparison: Indeterminate
     diff: -5.06%
     rme: ±6.97%

   normalize-document-wide
     Current:	 1,825 ops/sec (84 runs sampled)
     Reference:	 2,049 ops/sec (84 runs sampled)
     comparison: Indeterminate
     diff: -10.95%
     rme: ±6.00%

   render-split-block
     Current:	 43.32 ops/sec (116 runs sampled)
     Reference:	 39.68 ops/sec (110 runs sampled)
     comparison: Faster
     diff: +9.19%
     rme: ±5.31%

   split-block-deep
     Current:	 336 ops/sec (101 runs sampled)
     Reference:	 270 ops/sec (95 runs sampled)
     comparison: Faster
     diff: +24.11%
     rme: ±10.00%

   split-block-normal
     Current:	 559 ops/sec (96 runs sampled)
     Reference:	 495 ops/sec (99 runs sampled)
     comparison: Faster
     diff: +12.96%
     rme: ±6.73%

   split-block-wide
     Current:	 442 ops/sec (90 runs sampled)
     Reference:	 391 ops/sec (91 runs sampled)
     comparison: Faster
     diff: +13.14%
     rme: ±8.15%
```

I think the main benefit will be in react update.

Soreine commented on src/models/node.js:
```diff
@@ -163,18 +174,35 @@ const Node = {
+  forEachDescendant(iterator) {
+    return this.nodes.forEach((child, i, nodes) => {
+      if (iterator(child, i, nodes) === false) {
+        return false
```
Hmm so we can choose to skip a node's children by returning false ?
Where do you use it ?

I also noticed that `findDescendantDeep` loops over the nodes in the same order. So we could use `findDescendantDeep` wherever we used `forEachDescendant`.

Here is the looping orders of the three methods :

```

findDescendant

     *
   /   \          
  1     2         
 / \   / \        
3   4 5   6       
                  
                  
        
forEachDescendant,
findDescendantDeep
                         .
     *  
   /   \
  3     6  
 / \   / \ 
1   2 4   5
```

SamyPesse replied:
Returning `false` stop the iteration, I've commited a change to use `forEachDescendant` in `findDescendantDeep`
